Address Indexing (Part 2)

noescape00 opened the pull request:
Indexing is implemented in this PR. 

In the next one I'll implement public interface for address indexer, API methods, tests.

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+
+                this.dataStore = this.db.GetCollection<AddressIndexerData>(DbKey);
+
+                this.data = this.dataStore.FindAll().FirstOrDefault();
```
`this.data` can surely have a better name (same goes for `this.dataStore` its not clear what data store is we don't need a comment just a good variable name)

noescape00 replied:
fixed

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+
+namespace Stratis.Bitcoin.AddressIndexing
+{
+    public class AddressIndexer : IDisposable
```
A comment on the class about whit this component is doing at a high level is missing.

Also I think the `AddressIndexer ` needs to go either to its own feature or in the `BlockStoreFeature`.

We should try to abstract implementations away from the core code as much as possible.

noescape00 replied:
fixed

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+        {
+            if (this.nodeSettings.TxIndex)
+            {
+                string dbPath = Path.Combine(this.dataFolder.RootPath, "addressindex.litedb");
```
Brilliant and simple

noescape00 replied:
=)

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+            // Make sure it's on top of the tip.
+            bool blockAdded = item.Key;
+
+            if ((blockAdded && item.Value.ChainedHeader.Header.HashPrevBlock.ToString() != this.data.TipHash) ||
```
`.ToString()` ouch why?

noescape00 replied:
Because serializer that litedb uses can't work with uint256

dangershony replied:
:( won't it be then better to use `byte[]`?

noescape00 replied:
Fair point. Will do it in the next PR

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+            if ((blockAdded && item.Value.ChainedHeader.Header.HashPrevBlock.ToString() != this.data.TipHash) ||
+                (!blockAdded && item.Value.ChainedHeader.HashBlock.ToString() != this.data.TipHash))
+            {
+                const string message = "TransactionIndexer is in inconsistent state. This can happen if you've enabled txindex on an already synced or partially synced node. " +
```
So `txindex` also enables `addressindex`? is this correct? this is even more of a reason why this should be its own feature or part of store.

noescape00 replied:
Correct

dangershony replied:
I think there should be a separate flag for addressindex otherwise just by enabling trxindex you take so much more space.
Do we have stats of te size of hte db on STRAT  mainchain?

noescape00 replied:
>Do we have stats of te size of hte db on STRAT mainchain?

No. Just indexing the txes takes a while so I never synced the full chain. 
I will though.

noescape00 replied:
>I think there should be a separate flag for addressindex otherwise just by enabling trxindex you take so much more space.

Agreed. I'll add in the next PR

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+
+                this.logger.LogCritical(message);
+                this.logger.LogTrace("(-)[INCONSISTENT_STATE]");
+                throw new Exception(message);
```
There is a specific exception we can throw I beleive.

noescape00 replied:
Why? It's not supposed to be handled automatically

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+        /// <summary>Processes block that was added or removed from consensus chain.</summary>
+        private void ProcessBlockAddedOrRemoved(KeyValuePair<bool, ChainedHeaderBlock> item)
+        {
+            // Make sure it's on top of the tip.
```
Shouldnt the whole out of bound issue be dealt with at startup? (the Initialize method)

noescape00 replied:
Fair point, fixed

fassadlr commented on src/Stratis.Bitcoin.Features.BlockStore/BlockRepository.cs:
```diff
@@ -158,7 +158,10 @@ public Transaction GetTransactionById(uint256 trxid)
 
             if (!this.TxIndex)
+            {
+                this.logger.LogTrace("(-)[NOT_INDEXED]:null");
```
I would change this to `TX_INDEXING_DISABLED` or `TX_INDEX_OFF` then its immediately more clear.

noescape00 replied:
fixed

fassadlr commented on src/Stratis.Bitcoin.Features.BlockStore/BlockRepository.cs:
```diff
@@ -187,7 +190,44 @@ public Transaction GetTransactionById(uint256 trxid)
-            throw new NotImplementedException();
+            if (!this.TxIndex)
+            {
+                this.logger.LogTrace("(-)[NOT_INDEXED]:null");
```
Same here.

noescape00 replied:
fixed

fassadlr commented on src/Stratis.Bitcoin.Features.BlockStore/BlockRepository.cs:
```diff
@@ -187,7 +190,44 @@ public Transaction GetTransactionById(uint256 trxid)
+                    if (!blockRow.Exists)
+                    {
+                        this.logger.LogTrace("(-)[NO_BLOCK]:null");
+                        return null;
```
Shouldn't this be a `continue`?  If the transaction doesn't exist, then just log and continue onto the next one?

noescape00 replied:
No, it's a sign of a mistake. If requested tx is not found smth is wrong. 
Components should never ask for a transaction that is not a part of consensus chain

fassadlr commented on src/Stratis.Bitcoin.Features.BlockStore/BlockRepository.cs:
```diff
@@ -187,7 +190,44 @@ public Transaction GetTransactionById(uint256 trxid)
+                    if (!transactionRow.Exists)
+                    {
+                        this.logger.LogTrace("(-)[NO_TX_ROW]:null");
+                        return null;
```
Rather `continue`?

fassadlr commented on src/Stratis.Bitcoin.Features.BlockStore/BlockRepository.cs:
```diff
@@ -187,7 +190,44 @@ public Transaction GetTransactionById(uint256 trxid)
+
+                    txes[i] = tx;
+                }
+
```
Extra line.

noescape00 replied:
fixed

fassadlr commented on src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs:
```diff
@@ -213,7 +213,55 @@ public Transaction GetTransactionById(uint256 trxid)
-            throw new NotImplementedException();
+            // Only look for transactions if they're indexed.
+            if (!this.nodeSettings.TxIndex)
+                return null;
```
Exit log here?

noescape00 replied:
fixed

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+                addrData = new AddressIndexData()
+                {
+                    ScriptPubKeyBytes = scriptPubKeyBytes,
+                    Changes = new List<AddressBalanceChange>()
```
Is it not cheaper interms of mempry to just track a balance?

dangershony replied:
Well actually this way is probably better as it makes it easier to track the height of every deposit

noescape00 replied:
it's cheaper but exchanges need a method that accepts minconf value so we need to know height of each change

dangershony commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexerData.cs:
```diff
@@ -0,0 +1,38 @@
+﻿using System.Collections.Generic;
+
+namespace Stratis.Bitcoin.AddressIndexing
```
Maybe rename this file to be `AddressIndexerTypes`?

noescape00 replied:
fixed

fassadlr commented on src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs:
```diff
@@ -213,7 +213,55 @@ public Transaction GetTransactionById(uint256 trxid)
+                {
+                    uint256 txId = trxids[i];
+
+                    foreach (ChainedHeaderBlock chainedHeaderBlock in this.pendingBlocksCache.Values)
```
Can't this foreach be replaced with :
```
this.pendingBlocksCache.Values.SelectMany(b=>b.Blocks).SelectMany(b=>b.Transactions)
```

Then you can just FirstOrDefault check if the tx exists?

noescape00 replied:
Fixed

fassadlr commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,284 @@
+                (!blockAdded && item.Value.ChainedHeader.HashBlock.ToString() != this.data.TipHash))
+            {
+                const string message = "TransactionIndexer is in inconsistent state. This can happen if you've enabled txindex on an already synced or partially synced node. " +
+                                       "Remove everything from the data folder and run the node with -txindex=true.";
```
That message should be -txindex=1

fassadlr commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexerData.cs:
```diff
@@ -0,0 +1,38 @@
+    public class AddressBalanceChange
+    {
+        /// <summary><c>true</c> if there was a deposit to an address, <c>false</c> if it was a withdrawal.</summary>
+        public bool Deposited { get; set; }
```
@noescape00 remember the changes I suggested here? I would rather change this an enum and have values Deposit and Withdrawal so that it is more explicit

noescape00 replied:
Bool will take less space during deserialization. Enum will consume whole integer or short

dangershony replied:
I am also fine wth a bool, much simpler.

fassadlr commented on src/Stratis.Bitcoin/AddressIndexing/AddressIndexerData.cs:
```diff
@@ -0,0 +1,38 @@
+        public long Satoshi { get; set; }
+
+        /// <summary>Height of a block in which operation was confirmed.</summary>
+        public int Height { get; set; }
```
Also this, rename to something like HeightBalanceChanged or something more descriptive.

noescape00 replied:
done

dangershony commented:
This PR make me think we should pull trx index out of store in to its own feature.

noescape00 commented:
@dangershony @fassadlr fixes applied

dangershony commented on src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,291 @@
+                // Subscribe to events.
+                this.blockConnectedSubscription = this.signals.Subscribe<BlockConnected>(blockConnectedData =>
+                {
+                    while (this.blockReceivedQueue.Count > 100)
```
This can be prevented by fetching a block if `ChainedHeaderBlock.Block` is empty in the dequeue delegate.

noescape00 replied:
Not sure what you mean. Could you elaborate? 

I wanted to stop consensus here to prevent bloating RAM with blocks that are not processed

dangershony replied:
If you fetch a block from disk it wont get cached, so if indexer is falling behind it can start fetching from disk and slowly catch up.

noescape00 replied:
Should it do it like that though? 

It will allow consensus sync faster but increase hard drive load. 

Also this code is almost never executed, txindexing itself is a bigger bottleneck

dangershony replied:
ok fine with me, if we find its actually a bottle neck then we address the issue 👍

dangershony commented on src/Stratis.Bitcoin/Base/BaseFeature.cs:
```diff
@@ -336,6 +336,8 @@ public override void Dispose()
             this.logger.LogInformation("Disposing finalized block info repository.");
             this.finalizedBlockInfoRepository.Dispose();
 
+            this.logger.LogInformation("Disposing address indexer.");
```
Delete this log

noescape00 replied:
I'll need it while syncing mainchain and then shutting the node down to see if it takes too long to save addresses

dangershony replied:
ok just remember to remove it later

dangershony commented on src/Stratis.Bitcoin.Features.BlockStore/BlockStoreFeature.cs:
```diff
@@ -146,6 +151,8 @@ public override async Task InitializeAsync()
 
             this.blockStoreSignaled.Initialize();
+
+            this.addressIndexer.Initialize();
```
I stil think we should consider two flags `-txindex` and `-addrindex` this component may be heavy

noescape00 replied:
yeah, I have it in my TODO list for the next PR

noescape00 replied:
this component is pretty big so I'm doing it in small chunks. there will be at least 2 more PRs

dangershony replied:
>yeah, I have it in my TODO list for the next PR
Excellent, though won't this mess up current running nodes?

dangershony replied:
Maybe you want to disable the indexer for now.

fassadlr replied:
Is it not disabled by default?

noescape00 replied:
It is disabled by default.

noescape00 commented:
@dangershony @fassadlr I've merged latest master in, moved txindex flags back to store settings and added additional flag that enables address indexing.

dangershony commented on src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexer.cs:
```diff
@@ -0,0 +1,298 @@
+                    Thread.Sleep(5000);
+                }
+
+                this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(true, blockConnectedData.ConnectedBlock));
```
I wonder if you may want to take the Block instance out here, it will disappear from `blockConnectedData` if more then 100 blocks behind consensus (alternatively you can fetch it form disk later) ok to do it in a following PR.

dangershony commented on src/Stratis.Bitcoin.Features.BlockStore/StoreSettings.cs:
```diff
@@ -23,6 +23,16 @@ public class StoreSettings
         public int MaxCacheSize { get; set; }
 
+        /// <summary><c>true</c> to maintain a full transaction index.</summary>
+        public bool TxIndex { get; set; }
```
What it was not in store settings before? hmm...

noescape00 replied:
It was, then moved to node settings, then back to BS settings again because you've convinced me to put indexer in BS feature

dangershony commented on src/Stratis.Bitcoin.Features.BlockStore/StoreSettings.cs:
```diff
@@ -23,6 +23,16 @@ public class StoreSettings
+
+        /// <summary><c>true</c> to maintain a full addresses index.</summary>
+        /// <remarks><see cref="TxIndex"/> should be set to <c>true</c> in order to enable address indexing.</remarks>
+        public bool IndexAddresses { get; set; }
```
For consistency `AddressIndex`?

noescape00 replied:
kk, will fix
